library/alloc/src/collections/btree/navigate.rs

```diff
@@ -232,6 +232,57 @@ impl<BorrowType, K, V> NodeRef<BorrowType, K, V, marker::LeafOrInternal> {
     }
 }
 
+impl<'a, K: 'a, V: 'a> Handle<NodeRef<marker::Immut<'a>, K, V, marker::Internal>, marker::Edge> {
+    /// Returns the leftmost internal edge just above the leaf level, that is underneath
+    /// an edge or is that edge itself.
+    fn first_height_1_edge(self) -> Self {
+        let mut edge = self;
+        while let Internal(internal) = edge.descend().force() {
+            edge = internal.first_edge();
+        }
+        edge
+    }
+}
+
+impl<'a, K: 'a, V: 'a> NodeRef<marker::Immut<'a>, K, V, marker::Internal> {
+    /// Returns the first edge for a depth first ordered visit of all internal edges.
+    pub fn dfs_first_internal_edge(
+        self,
+    ) -> Handle<NodeRef<marker::Immut<'a>, K, V, marker::Internal>, marker::Edge> {
+        self.first_edge().first_height_1_edge()
+    }
+}
+
+impl<'a, K: 'a, V: 'a> Handle<NodeRef<marker::Immut<'a>, K, V, marker::Internal>, marker::KV> {
+    /// Returns the next edge for a depth first ordered visit of all internal edges,
+    /// given the `right_kv` of the previous edge or an ancestor.
+    pub fn dfs_next_internal_edge(
+        self,
+    ) -> Handle<NodeRef<marker::Immut<'a>, K, V, marker::Internal>, marker::Edge> {
+        self.right_edge().first_height_1_edge()
+    }
+}
+
+impl<'a, K: 'a, V: 'a> NodeRef<marker::Immut<'a>, K, V, marker::LeafOrInternal> {
+    /// Visits all nodes in depth first order.
+    pub fn depth_first_search_nodes<F: FnMut(Self)>(self, mut visit: F) {
+        if let Internal(node) = self.force() {
+            let mut edge = node.dfs_first_internal_edge();
+            loop {
+                visit(edge.descend());
+                edge = match edge.right_kv() {
+                    Ok(kv) => kv.dfs_next_internal_edge(),
+                    Err(last_edge) => match last_edge.into_node().ascend() {
+                        Ok(parent_edge) => parent_edge,
+                        Err(_root) => break,
+                    },
+                };
+            }
+        }
+        visit(self);
+    }
```

Comment:
I am not 100% confident that we got this right -- in particular, the `dfs_{first,next}_internal_edge()` looks a bit suspicious to me -- maybe I'm just not following what you mean by DFS, though.

Pulling a [random tree image](http://www.btechsmartclass.com/data_structures/ds_images/Tree.png) from the web, I would expect the code to visit A, B, D, E, I, J, F, C, G, K, H. I think the current code will visit these differently though, starting at D -- that seems unexpected or at least not what I would call DFS.

That said without writing out a sample I'm not sure I can be confident, so I think it'd be helpful to have some tests written up for this DFS -- ideally we'd literally check the order we visited the nodes in, by pushing onto a vector for example. Not sure how I would identify nodes and such though.

Reply:
I didn't think about the terminology. Some wikipedia page says it is [the LNR variation of depth first search](https://en.wikipedia.org/wiki/Tree_traversal#In-order_(LNR)). But "in-order" seems a better name to me. Or "B-tree order"?

Obviously order matters only in one of the 3 places: in `dump_keys`, and we really want in-order there. It's also the same order as the rest of navigation.

Reply:
Regardless of that, I'm also not quite happy with a bunch of vaguely related functions. I'll try to come up with something better.

Reply:
Aha, yeah, in order and a link to that wikipedia page seems good. Would be great to test it, though.

Reply:
I didn't look hard but didn't find anything particular for B-tree, so ended up with a more intuitive (I hope) and equally reusable hybrid ordering.